Blog post "Design Patterns in Financial Data Science" (#22)
ImperialCollegeLondon/RSEBlog: docs/posts/DesignPatterns.md

---
date: 2024-07-24
author:
  - HarmonicReflux
categories:
  - Tutorial
tags:
  - Design Patterns / Gang of Four
  - Object Oriented Programming (OOP) / Functional Programming
  - Python
  - Financial Data Science
  - Software Engineering principles
---

# Understanding software design patterns in Financial Data Science

`Design Patterns` were first introduced in a book titled [Design Patterns: Elements of Reusable Object-Oriented Software](https://www.amazon.com/Design-Patterns-Elements-Reusable-Object-Oriented/dp/0201633612)
, published in 1994 by Erich Gamma, Richard Helm, Ralph Johnson and John Vissides, also known as the `Gang of Four`.
The book lies at the basis of Object-Oriented design theory and practice. A `Design Pattern` is a general, reusable solution to a commonly occurring problem.
One does not just find them in software development, but throughout engineering practice, as

> their philosophy is to find standard solutions to common problems in a given field, say, the `Schmidt coupling` in mechanical engineering, see also
>  [20 Mechanical Principles combined in a Useless Lego Machine" on YouTube [sic!]](https://www.youtube.com/watch?v=M1-YeqGynlw)

Reviewer: ```suggestion
their philosophy is to find standard solutions to common problems in a given field.
```

I'm not sure this adds anything, but instead detracts the reader from your post very early on.

> There are, also standard designs for car engines, for instance the four-stroke engine, or the shape of a kitchen sink in plumbing and house design.

Reviewer: ```suggestion
There are also standard designs for car engines, for instance the four-stroke engine, or the shape of a kitchen sink in plumbing and house design.
```

This is exactly what design patterns are for software. Each pattern solves a particular problem.
Some of these patterns focus in object creation, some of them focus on testing structure, say a system of objects, and some of them focus on how
objects should communicate.

There is only one problem: The book referenced above is about 30 years old and much has changed in the computing world, including the evolution of
programming languages. "Move fast and fail fast" mentality of start-ups, or, often occurring in the quantitative analysis world in finance

> the "focus on the question at hand to answer a particular problem, while caring to a lesser degree, or not at all, on the design of the program".

Reviewer: This sentence lacks structure

> Many Jupyter notebooks that are written in isolation, are a prime example of that. In that sense many data scientists are stuck on the "make it work"

Reviewer: ```suggestion
Many Jupyter notebooks that are written in isolation are a prime example of that. In that sense many data scientists are stuck on the "make it work"
```

> mantra, rather than progressing to "make it right", and, finally, "make it fast". That is not meant as a criticism, as more often than not, we rework, rewrite or re-manufacture entire parts of our code to achieve the "make it work" step and over-designing a piece of code can be seen as an issue in itself, at least at the prototyping, or "prove of concept" stage. For more mature systems this, however, leads to plenty of technical issues, which are difficult or even impossible to maintain.

Reviewer: ```suggestion
mantra, rather than progressing to "make it right", and, finally, "make it fast". That is not meant as a criticism, as more often than not we rework, rewrite or re-manufacture entire parts of our code to achieve the "make it work" step, and over-engineering a piece of code can be seen as an issue in itself, at least at the prototyping, or "proof-of-concept" stage. For more mature systems, however, this leads to plenty of technical issues, which are difficult or even impossible to maintain.
```

Also, `Object-Oriented programming` is not a prime style of writing software anymore, and many modern language have solid `functional programming`
support natively supported. One of such as language is `Python`, where functions are considered first-class citizens, i.e. we can pass them around

> like other objects and even make functions take other functions as input parameters.

Reviewer: I think this is potentially misleading, so it would be good to get other opinions! Python is first and foremost an OOP language, where everything (including functions) is an object.

Reviewer: Especially given that your python code at the end is object-oriented.

> This blog is aimed at providing an overview of design patterns that one can make use for in `Financial Data Science` applications.

Reviewer: ```suggestion
This article is aimed at providing an overview of design patterns that one can make use of in `Financial Data Science` applications.
```

> It does not claim to be exhaustive, and you are encouraged to take a look at the original publication, or even, to take the

Reviewer: ```suggestion
It does not claim to be exhaustive, and you are encouraged to take a look at the original publication, or even to take the
```

concepts provided and to implement them in another language of your choice.

<!-- more -->

## A necessary detour: Unified Modeling Language (UML)

Before we delve into the realm of Design Patterns, we should first describe how we can define objects and properties on the
drawing board, similarly as an architect first works on a blueprint before manifesting their ideas in the real world.
In software architecture one such a tool is the `Unified Modeling Language (UML)`. It is a standardised visual modeling
language used in the field of software engineering to specify, visualize, construct, and document the artifacts of software systems.

> UML is not a programming language but rather a tool used to help software developers visualise the design of a system.

Reviewer: Mix of visualise/visualize

### How does UML help us in Software Engineering?

UML provides software architects and designers with a language for developing clear models of system design. It includes a set of graphic notation techniques to create visual models of object-oriented software systems.
As a standardised language, UML makes it easier for individuals to understand the software design, whether they are part of the development team or external stakeholders. This helps in both collaborative settings and when transitioning projects between teams.

> Moreover, UML diagrams serve as a part of the documentation process, providing several types of diagrams that cater to different aspects of a software system. For example, use case diagrams for requirements gathering, class diagrams for showing static structure, and sequence diagrams for detailing the interactions.

Reviewer: ```suggestion
Moreover, UML diagrams serve as a part of the documentation process, providing several types of diagrams that cater to different aspects of a software system. For example, use-case diagrams for requirements gathering, class diagrams for showing static structure, and sequence diagrams for detailing the interactions.
```

By using UML, software developers can think through the design and discover potential issues before any code is actually written. This can save time and resources in the development process.
Also, UML helps break down complex systems into manageable components, which can be analysed, designed, and implemented independently.
Another strength of UML is that it is platform-independent. This means the same modeling language can be used to design software systems across various platforms.

### Types of UML diagrams

UML includes several types of diagrams, each serving different purposes:

- Structural diagrams (e.g., Class Diagrams, Object Diagrams): Focus on the static aspects of the system.

- Behavioural diagrams (e.g., Use Case Diagrams, Activity Diagrams): Focus on dynamic aspects.

- Interaction diagrams (part of behavioral diagrams, e.g., Sequence Diagrams, Communication Diagrams): Focus specifically on the flow of control and data among the elements in the system.

> UML hence is an important tool used across the software development lifecycle, from conceptualisation through to implementation. It helps teams communicate ideas more effectively and builds a foundation for creating a robust architecture. By using UML, developers can ensure that their initial designs can be translated into functional software, making it an important part of modern software engineering practices.

Reviewer: ```suggestion
Hence UML is an important tool used across the software development lifecycle, from conceptualisation through to implementation. It helps teams communicate ideas more effectively and builds a foundation for creating a robust architecture. By using UML, developers can ensure that their initial designs can be translated into functional software, making it an important part of modern software engineering practices.
```

### UML within Jupyter notebooks

> Form within Jupyter notebooks, we can make use of a library called `iplantuml` that we can install via pip upon invoking pip install iplantuml.

Reviewer: ```suggestion
From within Jupyter notebooks, we can make use of a library called `iplantuml` that we can install via pip upon invoking `pip install iplantuml`.
```

> You can also use it to lay out the structure of your software using the package and use that as a basis for a research report or a dissertation. It is in general a good idea to keep a record of the blueprit of your software, specifically if you implement larger works or production systems.

Reviewer: ```suggestion
You can also use it to lay out the structure of your software using the package and use that as a basis for a research report or a dissertation. It is in general a good idea to keep a record of the blueprint of your software, especially if you implement larger works or production systems.
```

        return d1, d2

    # method to compute the option price using the provided strategy
    def compute_option_price(self, strategy: OptionPriceStrategy) -> float:
        return strategy.compute_price(self)

# main function to demonstrate the usage of the Strategy Pattern
def main() -> None:
    # create instances of the pricing strategies
    call_strategy: OptionPriceStrategy = EuropeanCallStrategy()
    put_strategy: OptionPriceStrategy = EuropeanPutStrategy()

    # create a Black-Scholes option with given parameters
    option: BlackScholesOption = BlackScholesOption(
        stock_price=100,
        strike_price=100,
        time_to_maturity=1,
        risk_free_rate=0.05,
        volatility=0.2
    )

    # compute and print the call option price using the call strategy
    call_price: float = option.compute_option_price(call_strategy)
    print(f"Call price: {call_price:.2f} [monetary units].")

    # compute and print the put option price using the put strategy
    put_price: float = option.compute_option_price(put_strategy)
    print(f"Put price: {put_price:.2f} [monetary units].")

if __name__ == "__main__":
    main()
```

> In the currency of the contract, under the Black-Scholes model, the price of a European call and put options given the underlying stock are hence :

Reviewer: ```suggestion
In the currency of the contract, under the Black-Scholes model, the price of a European call and put options given the underlying stock are hence:
```

- Call price: 10.45 [monetary units].
- Put price: 5.57 [monetary units].

Note the separation of concerns:  `EuropeanCallStrategy()` knows nothing about `EuropeanPutStrategy()` and vice versa.
It is hence straight-forward to maintain and extend one part of Strategy without touching the other part.

## Conclusion

> In this blog, we introduced design patterns and demonstrated how to implement the `Strategy Pattern` using the Object-Oriented paradigm in Python.

Reviewer: ```suggestion
In this article, we introduced design patterns and demonstrated how to implement the `Strategy Pattern` using the Object-Oriented paradigm in Python.
```
